src/_pytest/python.py

```diff
@@ -911,6 +911,124 @@ def hasnew(obj: object) -> bool:
     return False
 
 
+@final
+@attr.s(frozen=True, auto_attribs=True)
```

Comment:
Can add `slots=True` here as well, doesn't hurt.

Reply:
> doesn't hurt

IIUC [attrs docs](https://www.attrs.org/en/stable/how-does-it-work.html#summary) correctly, `@attr.s(frozen=True, slots=True, ...)` actually could hurt in some cases:
> You should avoid instantiating lots of frozen slotted classes (i.e. @frozen) in performance-critical code

I can't reason offhand about whether we'd be affected, can you?

Reply:
Hmm interesting. We use this all over the place so we should probably check if it has any effect separately. But until then let's keep it in this PR.